core/backend/src/test/ecdb/ECSqlStatement.test.ts

```diff
@@ -3338,4 +3338,108 @@ describe("ECSqlStatement", () => {
 
     await assert.isRejected(reader.toArray(), "Struct type binding not supported");
   });
+
+  it("invalid relClassId in navigation properties", async () => {
+    using ecdb = ECDbTestHelper.createECDb(outDir, "bindnavigation.ecdb",
+      `<ECSchema schemaName="Test" alias="test" version="01.00.00" xmlns="http://www.bentley.com/schemas/Bentley.ECXML.3.1">
+        <ECEntityClass typeName="Parent" modifier="Sealed">
+          <ECProperty propertyName="Code" typeName="string"/>
+        </ECEntityClass>
+        <ECEntityClass typeName="Child" modifier="Sealed">
+          <ECProperty propertyName="Name" typeName="string"/>
+          <ECNavigationProperty propertyName="Parent" relationshipName="ParentHasChildren" direction="backward"/>
+        </ECEntityClass>
+        <ECRelationshipClass typeName="ParentHasChildren" modifier="None" strength="embedding">
+          <Source multiplicity="(0..1)" roleLabel="has" polymorphic="false">
+              <Class class="Parent"/>
+          </Source>
+          <Target multiplicity="(0..*)" roleLabel="has" polymorphic="false">
+              <Class class="Child"/>
+          </Target>
+        </ECRelationshipClass>
+      </ECSchema>`);
+
+    assert.isTrue(ecdb.isOpen);
+
+    let reader = ecdb.createQueryReader("SELECT ECInstanceId FROM meta.ECClassDef WHERE Name='Parent'");
+    assert.isTrue(await reader.step());
+    const parentClassId: Id64String = reader.current.ECInstanceId;
+    assert.isTrue(Id64.isValidId64(parentClassId));
+
+    // When the ECSql insert validation is set to true, the invalid relClassId is detected and an error is thrown.
+    reader = ecdb.createQueryReader("PRAGMA validate_ecsql_inserts=true");
+    assert.isTrue(await reader.step());
+
+    try {
+      ecdb.withCachedWriteStatement(`INSERT INTO test.Child(Parent.Id, Parent.RelECClassId) VALUES(1, ${parentClassId})`, () => {});
```

Comment:
Just pasting this as a comment from teams discussion
I think it would be good to have another test that uses insertElement and updateElement instead of a low level sql statement. Not sure if you can force an invalid class id in those though..
To make sure those methods fail gracefully if bind fails, as expected.